cc/build_defs.bzl

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright 2022 The Project Oak Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Provides a rule that outputs a monolithic static library."""
+
+load(
+    "@bazel_tools//tools/cpp:toolchain_utils.bzl",
+    "find_cpp_toolchain",
+    "use_cpp_toolchain",
+)
+
+# Bazel rule for collecting the header files that a target depends on.
+def _cc_static_library_impl(ctx):
+    output_lib = ctx.actions.declare_file("lib{}.a".format(ctx.attr.name))
+    output_flags = ctx.actions.declare_file("lib{}.link".format(ctx.attr.name))
+
+    cc_toolchain = find_cpp_toolchain(ctx)
+
+    # Aggregate linker inputs of all dependencies
+    lib_sets = []
+    for dep in ctx.attr.deps:
+        lib_sets.append(dep[CcInfo].linking_context.linker_inputs)
+    input_depset = depset(transitive = lib_sets)
+
+    # Collect user link flags and make sure they are unique
+    unique_flags = {}
+    for inp in input_depset.to_list():
+        unique_flags.update({
+            flag: None
+            for flag in inp.user_link_flags
+        })
+    link_flags = unique_flags.keys()
+
+    # Collect static libraries
+    libs = []
+    for inp in input_depset.to_list():
+        for lib in inp.libraries:
+            if lib.pic_static_library:
+                libs.append(lib.pic_static_library)
+            elif lib.static_library:
+                libs.append(lib.static_library)
+
+    script_file = ctx.actions.declare_file("lib{}.mri".format(ctx.attr.name))
+    commands = ["create {}".format(output_lib.path)]
+    for lib in libs:
+        commands.append("addlib {}".format(lib.path))
+    commands.append("save")
+    commands.append("end")
+    ctx.actions.write(
+        output = script_file,
```

Comment:
In our CI we always check that we don't generate new files in the repo.
Does this rule create a new file that is not tracked by the repository?

Reply:
These are bazel rules, I think their only output is within the bazel folders, they cannot possibly write to random parts of the host file system (I hope).

Reply:
Like what Tiziano said: the files (intermediates and final static .a lib) the rule generates are put in bazel-bin/**/* and not touching any of host files.